### PR TITLE
 Add a kafka sink for event and  fix #1344

### LINKED ISSUE #1344: Balance Data Across Partions in the Kafka Sink
All data arrived in the partition id 0.
The partition id is fixed to 0 in the code and the key is not used ;-(

Can we have a version that balance data on all partitions of a cluster Kafka ?

Is it possible ?

### LINKED ISSUE #1365: [Proposal]: Add a kafka sink for Event
Kafka is well suited for handling metrics. It might be useful for events as well. So I think it is meaningful to add a kafka sink for event. I have completed the development of kafka sink for event locally. Shall I commit a PR to accept some guys' comments?  Thanks!

### PR DESCRIPTION
**Pull request by miaoyq**
fix #1365 

Signed-off-by: yanqiang Miao <miao.yanqiang@zte.com.cn>

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/heapster/1369)

### CODE REVIEW

**Comment by miaoyq** [comment]
@piosz  PTAL  Thanks so much!

**Comment by huangyuqi** [comment]
I will review this PR @piosz

**Comment by miaoyq** [comment]
@huangyuqi   Thanks for review, If this PR has any problems, please let me know.

**Comment by piosz** [comment]
Thanks @huangyuqi for offering you help. Feel free to merge once it's ready.

**Comment by piosz** [comment]
@k8s-bot ok to test

**Comment by miaoyq** [comment]
@piosz  @huangyuqi  Could this PR be merged?  If this PR has any problems, please let me know. Thanks so much!

File: events/sinks/kafka/driver.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+	//create kafka producer
+	conf := kafka.NewProducerConf()
+	conf.RequiredAcks = proto.RequiredAcksLocal
+	sinkProducer := broker.Producer(conf)
+	glog.V(3).Infof("kafka sink setup successfully")
+	return sinkProducer, nil
+}
+
+func NewKafkaSink(uri *url.URL) (event_core.EventSink, error) {
```
Comment (huangyuqi) [inline events/sinks/kafka/driver.go]: moving this logic of initing action to common directory; https://github.com/kubernetes/heapster/tree/master/common

File: events/sinks/kafka/driver.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+func (sink *kafkaSink) Name() string {
+	return "Apache Kafka Sink"
+}
+
+func (sink *kafkaSink) Stop() {
+	// nothing needs to be done.
+}
+
+// setupProducer returns a producer of kafka server
+func setupProducer(sinkBrokerHosts []string, brokerConf kafka.BrokerConf) (kafka.Producer, error) {
```
Comment (huangyuqi) [inline events/sinks/kafka/driver.go]: This is also a common logic between metric and events sink;

File: events/sinks/kafka/driver.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+func (sink *kafkaSink) produceKafkaMessage(dataPoint KafkaSinkPoint, topic string) error {
+	start := time.Now()
+	jsonItems, err := json.Marshal(dataPoint)
+	if err != nil {
+		return fmt.Errorf("failed to transform the items to json : %s", err)
+	}
+	message := &proto.Message{Value: []byte(string(jsonItems))}
+	_, err = sink.producer.Produce(topic, partition, message)
```
Comment (huangyuqi) [inline events/sinks/kafka/driver.go]: there is a issue about the partition.
https://github.com/kubernetes/heapster/issues/1344
It should be fixed in this PR, in my opinion.

**Comment by miaoyq** [comment]
Thanks @huangyuqi  for review, I will fix the issue, and refactor the code based on your instructions.

**Comment by miaoyq** [comment]
I have refactored kafka sink and fixed #1344,  PTAL @huangyuqi

**Comment by huangyuqi** [comment]
LGTM, @k8s-bot ok to test
Thanks, @miaoyq , I have no doubt about this PR;

**Comment by miaoyq** [comment]
@piosz  Could this PR be merged?  PTAL, Thanks a lot!

**Comment by miaoyq** [comment]
@piosz  The PR has been open for a long time, but is still not merged. If there is any problem still, I will be happy to modify.   PTAL, Thanks a lot!

**Comment by piosz** [comment]
@miaoyq thanks for the contribution. Apologies for the delay.

**Comment by miaoyq** [comment]
@adelbot  Please refer to the contents [kafka](https://github.com/kubernetes/heapster/blob/master/docs/sink-configuration.md#kafka)

**Comment by miaoyq** [comment]
Apologies for my mistaken, now heapster have been divided into two parts, namely `heapster `(for metrics) and `eventer`(for events). 
we should use `eventer` to received events, look like:
```
./eventer --sink="kafka:?brokers=localhost:9092&brokers=localhost:9093&eventstopic=testtopic"  ... ...
```

I wil update the docs.

**Pull request merged**